lib/rust/ensogl/core/src/display/object.rs

```diff
@@ -37,39 +36,53 @@ pub mod traits {
 
 /// A wrapper around [`Instance`] with attached sublayers. It is used for
 /// dynamic creation and managing of additional on-demand sublayers needed for various UI
-/// components.
+/// components. Sublayers are layers attached to other layers by [`Layer::add_sublayer`] method
+/// and used for shapes ordering and masking. See [`Layer`] documentation for more info about
+/// sublayers.
 ///
-/// When the Instance is added to another layer, [`Sublayers::remove_from_layer`] and
-/// [`Sublayers::add_to_layer`] methods of the `sublayers` field will be called with previous and
-/// new layers respectively.
+/// See [`crate::display::scene::layer::MaskingSublayers`] docs for one of the use cases.
 ///
-/// See [`MaskingSublayers`] docs for one of the use cases.
-#[derive(Debug, Clone, CloneRef)]
-pub struct InstanceWithSublayers<S: CloneRef> {
+/// When the [`Instance`] is added to another layer, [`Sublayers::for_each_sublayer`] method of the
+/// `sublayers` field will be used to reattach sublayers to this new layer as well. Thus the
+/// sublayers are following the [`Instance`] at all times.
+#[derive(Debug)]
+pub struct InstanceWithSublayers<S> {
     inner:         Instance,
-    /// Attached sublayers. Should implement [`Sublayers`] trait.
+    /// Attached sublayers. Should implement [`ForEachSublayer`] trait.
     pub sublayers: S,
 }
 
+impl<S: Clone> Clone for InstanceWithSublayers<S> {
+    fn clone(&self) -> Self {
+        Self { inner: self.inner.clone_ref(), sublayers: self.sublayers.clone() }
+    }
+}
```

Comment:
I think the derive macro works exactly this way: it adds the requirement of each type to implement Clone (even if this does not make sense).

Reply:
Oh, indeed, I think I was confused by the compiler error from `CloneRef` derive.